File: examples/bun-app/src/handlers/data-pipeline.ts
/**
 * Data pipeline step handlers.
 *
 * 8 steps demonstrating a fan-out / fan-in DAG:
 *   Extract (3 parallel) -> Transform (3 parallel) -> Aggregate -> GenerateInsights
 *
 * Thin DSL wrappers that delegate to ../services/data-pipeline for business logic.
 */

import { defineHandler } from '@tasker-systems/tasker';
import type { DateRange } from '../services/types';
import * as svc from '../services/data-pipeline';

// ---------------------------------------------------------------------------
// Step 1: ExtractSalesData (parallel with steps 2 and 3)
// ---------------------------------------------------------------------------

export const ExtractSalesDataHandler = defineHandler(
  'DataPipeline.StepHandlers.ExtractSalesDataHandler',
  { inputs: { sources: 'sources', dateRange: 'date_range' } },
  async ({ sources, dateRange }) =>
    svc.extractSalesData(
      sources as string[] | undefined,
      dateRange as DateRange | undefined,
    ),
);

// ---------------------------------------------------------------------------
// Step 2: ExtractInventoryData (parallel with steps 1 and 3)
// ---------------------------------------------------------------------------

export const ExtractInventoryDataHandler = defineHandler(
  'DataPipeline.StepHandlers.ExtractInventoryDataHandler',
  { inputs: { parameters: 'parameters' } },
  async ({ parameters }) =>
    svc.extractInventoryData(
      parameters as Record<string, unknown> | undefined,
    ),
);

// ---------------------------------------------------------------------------
// Step 3: ExtractCustomerData (parallel with steps 1 and 2)
// ---------------------------------------------------------------------------

export const ExtractCustomerDataHandler = defineHandler(
  'DataPipeline.StepHandlers.ExtractCustomerDataHandler',
  { inputs: { dateRange: 'date_range' } },
  async ({ dateRange }) =>
    svc.extractCustomerData(dateRange as DateRange | undefined),
);

// ---------------------------------------------------------------------------
// Step 4: TransformSales (depends on ExtractSalesData)
// ---------------------------------------------------------------------------

export const TransformSalesHandler = defineHandler(
  'DataPipeline.StepHandlers.TransformSalesHandler',
  { depends: { extractResults: 'extract_sales_data' } },
  async ({ extractResults }) =>
    svc.transformSales(extractResults as Record<string, unknown>),
);

// ---------------------------------------------------------------------------
// Step 5: TransformInventory (depends on ExtractInventoryData)
// ---------------------------------------------------------------------------

export const TransformInventoryHandler = defineHandler(
  'DataPipeline.StepHandlers.TransformInventoryHandler',
  { depends: { extractResults: 'extract_inventory_data' } },
  async ({ extractResults }) =>
    svc.transformInventory(extractResults as Record<string, unknown>),
);

// ---------------------------------------------------------------------------
// Step 6: TransformCustomer (depends on ExtractCustomerData)
// ---------------------------------------------------------------------------

export const TransformCustomerHandler = defineHandler(
  'DataPipeline.StepHandlers.TransformCustomersHandler',
  { depends: { extractResults: 'extract_customer_data' } },
  async ({ extractResults }) =>
    svc.transformCustomers(extractResults as Record<string, unknown>),
);

// ---------------------------------------------------------------------------
// Step 7: AggregateData (depends on all 3 transforms)
// ---------------------------------------------------------------------------

export const AggregateDataHandler = defineHandler(
  'DataPipeline.StepHandlers.AggregateMetricsHandler',
  {
    depends: {
      salesData: 'transform_sales',
      inventoryData: 'transform_inventory',
      customerData: 'transform_customers',
    },
  },
  async ({ salesData, inventoryData, customerData }) =>
    svc.aggregateMetrics(
      salesData as Record<string, unknown> | undefined,
      inventoryData as Record<string, unknown> | undefined,
      customerData as Record<string, unknown> | undefined,
    ),
);

// ---------------------------------------------------------------------------
// Step 8: GenerateInsights (depends on AggregateData)
// ---------------------------------------------------------------------------

export const GenerateInsightsHandler = defineHandler(
  'DataPipeline.StepHandlers.GenerateInsightsHandler',
  { depends: { metrics: 'aggregate_metrics' } },
  async ({ metrics }) =>
    svc.generateInsights(metrics as Record<string, unknown>),
);

File: examples/fastapi-app/app/handlers/data_pipeline.py
"""Data pipeline analytics step handlers.

8 steps forming a DAG pattern with 3 parallel extract branches:
  extract_sales_data      ─┐
  extract_inventory_data  ─┤─> transform_sales     ─┐
  extract_customer_data   ─┘   transform_inventory  ─┤─> aggregate_metrics -> generate_insights
                                transform_customers  ─┘

Thin DSL wrappers that delegate to app.services.data_pipeline for business logic.
"""

from __future__ import annotations

from tasker_core.step_handler.functional import depends_on, inputs, step_handler
from tasker_core.types import StepContext

from app.services import data_pipeline as svc
from app.services.types import (
    DataPipelineInput,
    PipelineAggregateMetricsResult,
    PipelineExtractCustomerResult,
    PipelineExtractInventoryResult,
    PipelineExtractSalesResult,
    PipelineTransformCustomersResult,
    PipelineTransformInventoryResult,
    PipelineTransformSalesResult,
)


# ---------------------------------------------------------------------------
# Extract handlers
# ---------------------------------------------------------------------------


@step_handler("extract_sales_data")
@inputs(DataPipelineInput)
def extract_sales_data(inputs: DataPipelineInput, context: StepContext):
    return svc.extract_sales_data(
        source=inputs.source,
        date_range_start=inputs.date_range_start,
        date_range_end=inputs.date_range_end,
        granularity=inputs.granularity,
    )


@step_handler("extract_inventory_data")
@inputs(DataPipelineInput)
def extract_inventory_data(inputs: DataPipelineInput, context: StepContext):
    return svc.extract_inventory_data(
        source=inputs.source,
        date_range_start=inputs.date_range_start,
    )


@step_handler("extract_customer_data")
@inputs(DataPipelineInput)
def extract_customer_data(inputs: DataPipelineInput, context: StepContext):
    return svc.extract_customer_data(source=inputs.source)


# ---------------------------------------------------------------------------
# Transform handlers
# ---------------------------------------------------------------------------


@step_handler("transform_sales")
@depends_on(sales_data=("extract_sales_data", PipelineExtractSalesResult))
def transform_sales(sales_data: PipelineExtractSalesResult, context: StepContext):
    return svc.transform_sales(sales_data=sales_data)


@step_handler("transform_inventory")
@depends_on(traffic_data=("extract_inventory_data", PipelineExtractInventoryResult))
def transform_inventory(traffic_data: PipelineExtractInventoryResult, context: StepContext):
    return svc.transform_inventory(traffic_data=traffic_data)


@step_handler("transform_customers")
@depends_on(inventory_data=("extract_customer_data", PipelineExtractCustomerResult))
def transform_customers(inventory_data: PipelineExtractCustomerResult, context: StepContext):
    return svc.transform_customers(inventory_data=inventory_data)


# ---------------------------------------------------------------------------
# Aggregate and insight handlers
# ---------------------------------------------------------------------------


@step_handler("aggregate_metrics")
@depends_on(
    sales_transform=("transform_sales", PipelineTransformSalesResult),
    traffic_transform=("transform_inventory", PipelineTransformInventoryResult),
    inventory_transform=("transform_customers", PipelineTransformCustomersResult),
)
def aggregate_metrics(
    sales_transform: PipelineTransformSalesResult,
    traffic_transform: PipelineTransformInventoryResult,
    inventory_transform: PipelineTransformCustomersResult,
    context: StepContext,
):
    return svc.aggregate_metrics(
        sales_transform=sales_transform,
        traffic_transform=traffic_transform,
        inventory_transform=inventory_transform,
    )


@step_handler("generate_insights")
@depends_on(metrics=("aggregate_metrics", PipelineAggregateMetricsResult))
def generate_insights(metrics: PipelineAggregateMetricsResult, context: StepContext):
    return svc.generate_insights(metrics=metrics)
